docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,536 @@
+
+    ```go
+        // LateInitializeConfig contains instructions for how to handle the
+        // retrieval and setting of server-side defaulted fields.
+        type LateInitializationConfig struct {
+            // DelaySeconds provides the base delay to attempt late initialization again after an unsuccessful attempt
+            // to late initialized fields from ReadOne output
+            DelaySeconds int `json:"delay_seconds,omitempty"`
+            // MaxBackoffSeconds provide the maximum allowed backoff when retrying late initialization after an    
+            // unsuccessful attempt
+            MaxBackoffSeconds int `json:"max_backoff_seconds,omitempty"`
+            // MaxRetries provide the maximum number of retries for setting late initialized fields.
+            MaxRetries int `json:"max_retries,omitempty"`
+        }
+
+        type FieldConfig struct {
+            LateInitialize *LateInitializeConfig `json:"late_initialize,omitempty"`
+        }
+    ```
+
+3. **Provide 2 new hooks to AWS service teams to customize the code generation in step 4**
+    
+    These 2 hooks will be:
+
+    a) **late_initialize_pre_read_one**:- This hook will provide the custom code to be inserted **before** setting server side defaults for missing optional spec fields.
+
+    b) **late_initialize_post_read_one**: This hook will provide the custom code to be inserted **after** setting server side defaults for missing optional spec fields.
+
+4. Add `AWSResourceManager.LateInitialize()` method to handle server side defaults for fields mentioned in LateInitializationConfig.
+
+* Proposed Pseudo Code
+    ```go
+    // manager.go.tpl
+    func (rm *resourceManager) LateInitialize(
+	ctx context.Context,
+    latest acktypes.AWSResource,
+    ) (acktypes.AWSResource, error) {
+        {{- if $hookCode := Hook .CRD "late_initialize_pre_read_one" }}
+            {{ $hookCode }}
+        {{- end }}
+  
+        // 1. Filter all the fields which have LateInitializationConfig.
+        // 2. If there are no field needing late initialization, return  `latest, nil`
+        // 3. DeepCopy `latest` object into `latestWithDefaults`
+        // 4. Perform ReadOne operation and set late initialized fields in `latestWithDefaults` with nil checks
+        // 5. Collect the lateInitialized fields which are still not set in `latestWithDefault` into 
+        // `uninitializedFieldNames` slice.
+        // 6. If `uninitializedFieldNames` is not empty
+                  // 6.a. Using `services.k8s.aws/late-initialization-attempt` annotation, calculate the
+                  //    maximum delay needed for requeue amongst the `uninitializedFieldNames`
+                  // 6.b. Update the `LateInitialized` condition and `services.k8s.aws/late-initialization-attempt` annotation
+                  //    in latestWithDefaults object
+                  // 6.c. return latestWithDefaults, requeueNeededAfter(max-delay-from-6a)
```

Comment:
Sorry, how does this support retries? This seems to imply it will continue indefinitely - it never references `MaxRetries`.

Reply:
My miss. Adding it now.

Reply:
I thought about it and I do not think we need MaxRetries for late initialization config.

In Kubernetes, the resource should keep trying to sync if the desired and latest state of object do not match, unless there is a terminal condition.

In the context of late initialization, I believe there is no terminal condition when trying to set defaults using ReadOne output. Since the service teams have defined a certain field will be defaulted at some point of time when it is not originally provided by the user, then the reconciler should keep up the attempts to find defaulted value of the field.

Your suggestion about exponential backoff with max backoff will help prevent the retry storm for these async cases.

Reply:
That sounds solid. Exponential backoff while waiting for defaults, yep. It's also a one-way door should we ever need to implement that later down the road.